File: src/main/java/no/ntnu/webdev/webproject7/product/Product.java
package no.ntnu.webdev.webproject7.product;

import javax.persistence.Entity;
import javax.persistence.Id;

@Entity
public class Product {
    @Id
    private String id;
    private Double price;
    private Double discount;
    private String image;
    private String title;
    private String description;
    private String weight;

    public Product(String id, Double price, Double discount, String image, String title, String description, String weight) {
        this.id = id;
        this.price = price;
        this.discount = discount;
        this.image = image;
        this.title = title;
        this.description = description;
        this.weight = weight;
    }

    protected Product() {}

    public String getId() {
        return this.id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public Double getPrice() {
        return this.price;
    }

    public void setPrice(Double price) {
        this.price = price;
    }

    public Double getDiscount() {
        return this.discount;
    }

    public void setDiscount(Double discount) {
        this.discount = discount;
    }

    public String getImage() {
        return this.image;
    }

    public void setImage(String image) {
        this.image = image;
    }

    public String getTitle() {
        return this.title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public String getDescription() {
        return this.description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getWeight() {
        return this.weight;
    }

    public void setWeight(String weight) {
        this.weight = weight;
    }
}

File: src/main/java/no/ntnu/webdev/webproject7/product/Product.kt
package no.ntnu.webdev.webproject7.product

import no.ntnu.webdev.webproject7.comment.Comment
import no.ntnu.webdev.webproject7.crud.CrudModel
import no.ntnu.webdev.webproject7.utilities.objectsNotNull
import javax.persistence.*

typealias ProductId = Long

enum class Category(val type: String){
    TEA("tea"),
    ACCESSORIES("accessories")
}

enum class Subcategory(val type: String){
    BLACK_TEA("black tea"),
    GREEN_TEA("green tea"),
    WHITE_TEA("white tea"),
    CUP("cups")
}

@Entity
open class Product(
    @Column(nullable = true)
    @OneToMany(cascade = [CascadeType.ALL])
    val comments: List<Comment>? = null,

    @Column(nullable = false)
    var price: Double? = null,

    @Column(nullable = false)
    var discount: Double? = null,

    @Column(nullable = true)
    var image: String? = null,

    @Column(nullable = false)
    var title: String? = null,

    @Column(nullable = false)
    var description: String? = null,

    @Column(nullable = false)
    var weight: String? = null,

    @Column(nullable = false)
    val category: Category? = null,

    @Column(nullable = false)
    val subcategory: Subcategory? = null

) : CrudModel<ProductId> {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    override var id: ProductId = 0

    protected constructor() : this(null)

    override fun validate(): Boolean {
        return objectsNotNull(discount, price, title, description, weight, category, subcategory);
    }
}
